Rewrote test suite documentation (#137)
castep-docs/castep-docs: docs/developer/test_suite.md

The CASTEP source distribution includes a test suite to aid developers and users or HPC admins alike by

- Ensuring that new development and refactored code does not cause a crash or significantly change any results;
- Testing a new compiler, or new version of an existing compiler for failures due to language standard support, 
  non-portable code or bugs in the compiler itself;
- Ensuring that serial and parallel execution yield the same results;
- Automating the tests so they may be run as a routine part of
  development, either my the developer or routinely by the "Jenkins"
  Continuous Integration (CI) system.

We adopt the testing method known as *regression testing* whereby
CASTEP is run on a large set of input decks for small calculations,
and the result compared to a previously generated and known good
outputs, known as the *benchmark* for each test. The automation and
testing are performed using the
[`testcode`](https://testcode.readthedocs.io/en/latest/) testing
platform. This handles all of the definition of test categories, the
selection, sequencing and running of tests, and comparison of the run
output with a benchmark. To run CASTEP, postprocess the output and

> extract the comparison data it invokes two scripts supplied in

Reviewer: ```suggestion
extract the comparison data, it invokes two scripts supplied in
```

Author: OK

the CASTEP distribution `bin` directory named [`run_castep_test.py`](#creation-and-format-of-test-outputs-and-benchmarks) and
[`extract_results.pl`](#extraction-of-properties).

The directory structure of the suite consists of a top-level `Test` directory containing testcode's configuration files, a GNU Makefile which contains targets to drive the tests, and a set of subdirectories dividing the tests into categories; in outline

```
Test
 |-- CMakeLists.txt
 |-- Makefile
 |-- jobconfig
 |-- userconfig
 |-- Electronic
         |-- Cr-AFM
			 |-- Cr_00PBE.uspcc
			 |-- Cr_AFM.cell
			 |-- Cr_AFM.param
			 |-- benchmark.out.castep-23.1.castep-python-1.0.python-3.6.9.inp=Cr_AFM.param
         |-- <Electronic test 2> 
|-- <Category 2>
         |-- <Category 2 test 1>
         |-- <Category 2 test 2>
 ...
```

## Testing your code

### A complete test run

> A run of the test suite is started by running the testcode scipt `<CASTEP_ROOT>/bin/testcode.py` from within the Test didectory, but as some options are usually required, this is most conveniently invoked via the build system by either

Reviewer: ```suggestion
A run of the test suite is started by running the testcode script `<CASTEP_ROOT>/bin/testcode.py` from within the Test directory, but as some options are usually required, this is most conveniently invoked via the build system by either
```

Author: OK

```Shell
    cd Test
    make check
```

or

```Shell
    cmake -t check
```

following a compile with either the GNU make or cmake build systems. This will produce a brief summary output along the lines of

```Shell
make check COMMS_ARCH=mpi
Found mpirun, using PARALLEL_CMD=mpirun -np tc.nprocs
rm -f */*/*.castep */*/*.dfpt_wvfn */*/*.fd_wvfn */*/*.wvfn.* */*/*.*.err
../bin/testcode.py -q  --processors=4 --total-processors=36 --user-option castep run_cmd_template "../../../bin/run_castep_test.pl tc.program tc.args tc.input tc.output mpirun -np tc.nprocs" -e /home/user/CASTEP-23.1/bin/castep.mpi -c simple -c d3-simple -c d4-simple -c md-parallel
....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................... [487/487]
```

> if everything passes or with `F` replacing one or more of the `.` symbols to indicate a failed test or tests.

Reviewer: Does the test-suite have skip (`S`) at all?

Author: Missed that, thanks.

In case of a failed test, more detailed information on which tests failed and why, the commands 
`make compare`, `make diff`, `cmake -t compare`, or `cmake -t diff`
will print a summary or detailed analysis of the failed tests to stdout.

### Partial and incremental testing

During a code development process, running the entire test suite to
check every change will be too time consuming and functionally
unnecessary.  The tests are divided into categories, named `spe`
,`bs` ,`geom` ,`md` ,`magres` ,`misc`
,`pair-pot` ,`phonon` ,`otf` ,`pseudo`
,`tddft` ,`ts` ,`xc` ,`nlxc` ,`libxc`
,`solvation` ,`d3` ,`d4` ,`python`
and `parallel` corresponding roughly to the subdirectories

```
Test/
  |-Electronic
  |-Excitations
  |-Geometry
  |-MD
  |-Magres
  |-Misc
  |-Pair-Pot
  |-Phonon
  |-Pseudo
  |-TDDFT
  |-XC
  |-Solvation
  |-Dispersion
  |-Pseudopotentials
  |-Python
  |-Utilities
  |-LibXC
  |-Parallel
```

and may be run individually using

```
    make check-<category>
    make compare-<category>
    make diff-<category>
```

or the targets of the same name in cmake.

??? keyword "Running individual tests"

>     In fact the tests may be run at even finer granularity, down to the subdirectory. This is not implemented via the Makefile or cmake, and can only be accessed by invoking `testcode` directly.  The most straightforward means of doing this is to copy and paste the testcode invocation command printed by `make check-spe` including all of the run and argument setup and change the argument of the `-c` option to `<directory>/<subdirectory>`. See section [Invoking Testcode](#direct-invocation-of-testcode)

Reviewer: > The most straightforward means of doing this is to copy and paste the testcode invocation command printed by `make check-spe`

Provide it here instead? While potentially liable to change I doubt the command will change sufficiently frequently to not warrant putting it here. 

At the very least make it clear the link to below contains the requisite lines.

Author: Reworded for clarity

  * When adding is a major new capability, it may be appropriate to add a new top-level category and subdirectory tree. In that case the [`Makefile`](#adding-to-make) and [`CMakeLists.txt`](#adding-to-cmake) will need to be extended to add the new `check`, `compare`, `diff` and `recheck` targets.
  * CASTEP input files: `.cell` and `.param`
  * If any external pseudopotential files are required, these should be placed in the `Test/Pseudopotentials` subdirectory and a symbolic link made from the test subdirectory. Both the fle and the link should be added to the `git` repository. 
  * to modify [`bin/run_castep_test.py`](#test-outputs-and-benchmarks) if you need to test the contents of a new output file.
  * if testing a new quantity, either in the `.castep` file or any other output file, you must extend the [`bin/extract_results.pl`](#extraction-of-properties) script to scan for and report the new value to be tested.

>   * A new Stanza in [`jobconfig`](#jobconfig-file) will be required to specify tolerances for the comparison if the defaults in [`userconfig`](#userconfig-file) are not suitable.

Reviewer: Does `stanza` need capitalising here?

Author: no

Reviewer: ```suggestion
  * A new stanza in [`jobconfig`](#jobconfig-file) will be required to specify tolerances for the comparison if the defaults in [`userconfig`](#userconfig-file) are not suitable.
```

  * a new benchmark file containing CASTEP output from a known good
run. Unfortunately creating benchmark files still requires some manual
intervention.  Regrettably `testcode`s option to create a new
benchmark is unsuitable as it is too inflexible to create a single
test without requiring updates to the entire test suite.  The easiest
way to do this is

    a. Run testcode, either from Make or by hand.  The new test will
fail because no benchmark is present for comparison.

    b. Copy or rename the just created test file

> "test.out.<date>?inp=<seedname>.param    to
> "benchmark.out.castep-<old-castep-version>castep-python-1.0.python-<old-python-version>.inp=<seedname>.param".
> where "<old-castep-version>" and "<old-python-version>" should be chosen to match the existing "benchmark.out" files from other tests.

Reviewer: ```suggestion
"test.out.<date>?inp=<seedname>.param"  to
"benchmark.out.castep-<old-castep-version>castep-python-1.0.python-<old-python-version>.inp=<seedname>.param".
where "<old-castep-version>" and "<old-python-version>" should be chosen to match the existing "benchmark.out" files from other tests.
```
Double check this renders since HTML style `<>` can be interpreted by Markdown. You *may* need to escape them.

Author: I have changed the quotes so this renders correctly.
